### PR TITLE
chore(griffin) - various view and functions needed for Kafka metadata queries

### PR DESCRIPTION
**Pull request by jaugsburger**

### CODE REVIEW

File: core/src/test/java/io/questdb/griffin/engine/functions/catalogue/AttributeCatalogueFunctionFactoryTest.java [hunk 1]
```diff
@@ -262,17 +262,61 @@ public void testPgAttributeFuncWith2Tables() throws Exception {
...
+                "attrelid\tattname\tattnum\tatttypid\tattnotnull\tatttypmod\tattlen\tattidentity\tattisdropped\n" +
+                        "1\ta\t1\t0\tfalse\t0\t0\t\tfalse\n",
                 true,
                 false,
                 true
         );
     }
+
+    @Test
+    public void testSecondKafkaMetadataQuery() throws Exception {
```
Comment (bluestreak01) [inline core/src/test/java/io/questdb/griffin/engine/functions/catalogue/AttributeCatalogueFunctionFactoryTest.java:280]: the goal of the SQL is to enumerate table columns like PG does

```
 nspname | relname  |  attname   | atttypid | attnotnull | atttypmod | attlen | typtypmod | attnum | attidentity | adsrc | description | typbasetype | typtype
---------+----------+------------+----------+------------+-----------+--------+-----------+--------+-------------+-------+-------------+-------------+---------
 public  | po_items | po_no      |       23 | t          |        -1 |      4 |        -1 |      1 |             |       |             |           0 | b
 public  | po_items | item_no    |       23 | t          |        -1 |      4 |        -1 |      2 |             |       |             |           0 | b
 public  | po_items | product_no |       23 | f          |        -1 |      4 |        -1 |      3 |             |       |             |           0 | b
 public  | po_items | qty        |       23 | f          |        -1 |      4 |        -1 |      4 |             |       |             |           0 | b
 public  | po_items | net_price  |     1700 | f          |        -1 |     -1 |        -1 |      5 |             |       |             |           0 | b
(5 rows)
```

File: core/src/main/java/io/questdb/griffin/engine/functions/catalogue/AttributeCatalogueFunctionFactory.java [hunk 2]
```diff
@@ -236,6 +246,12 @@ public int getStrLen(int col) {
         metadata.add(new TableColumnMetadata("attrelid", ColumnType.INT, null));
         metadata.add(new TableColumnMetadata("attname", ColumnType.STRING, null));
         metadata.add(new TableColumnMetadata("attnum", ColumnType.SHORT, null));
+        metadata.add(new TableColumnMetadata("atttypid", ColumnType.INT, null));
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/functions/catalogue/AttributeCatalogueFunctionFactory.java:249]: the aim of the SQL is to enum columns. Having constant 0 out of the field isn't going to work

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

File: core/src/main/java/io/questdb/griffin/engine/functions/catalogue/DescriptionCatalogueFunctionFactory.java [hunk 3]
```diff
@@ -156,46 +158,86 @@ public long size() {
...
 
+            @Override
+            public short getShort(int col) {
+                return columnNumber;
+            }
+
             @Override
             public CharSequence getStr(int col) {
-                return "table";
+                return columnNumber == 0 ? "table" : "column";
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/functions/catalogue/DescriptionCatalogueFunctionFactory.java:240]: use sparse array

File: core/src/main/java/io/questdb/griffin/engine/functions/catalogue/PrefixedTypeCatalogueFunctionFactory.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+public class PrefixedTypeCatalogueFunctionFactory extends TypeCatalogueFunctionFactory {
+
+    @Override
+    public String getSignature() {
+        return "pg_catalog.pg_type()";
+    }
+
+    @Override
+    public Function newInstance(ObjList<Function> args, int position, CairoConfiguration configuration) {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/functions/catalogue/PrefixedTypeCatalogueFunctionFactory.java:41]: check if base type does the same and remove if duplicate

File: core/src/main/java/io/questdb/griffin/engine/functions/catalogue/TypeCatalogueCursor.java [hunk 5]
```diff
@@ -69,28 +62,51 @@ public void toTop() {
...
+        return rowCount;
     }
 
-    private class DescriptionCatalogueRecord implements Record {
-
+    private class TypeCatalogueRecord implements Record {
         @Override
         public int getInt(int col) {
-            return intColumns[col][row];
+            return col == 0 ? PG_TYPE_OIDS.get(row) : col == 4 ? PgOIDs.PG_PUBLIC_OID : 0;
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/functions/catalogue/TypeCatalogueCursor.java:71]: use sparse array

File: core/src/main/java/io/questdb/griffin/engine/functions/catalogue/TypeCatalogueCursor.java [hunk 6]
```diff
@@ -69,28 +62,51 @@ public void toTop() {
...
         }
 
         @Override
         public CharSequence getStr(int col) {
-            return descriptions[row];
+            return PG_TYPE_TO_NAME.get(PG_TYPE_OIDS.get(row));
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/functions/catalogue/TypeCatalogueCursor.java:76]: `PG_TYPE_TO_NAME` should be an array (dense)

File: core/src/main/java/io/questdb/griffin/engine/functions/catalogue/TypeCatalogueFunctionFactory.java [hunk 7]
```diff
@@ -24,32 +24,15 @@
...
 import io.questdb.griffin.FunctionFactory;
-import io.questdb.griffin.SqlExecutionContext;
-import io.questdb.griffin.engine.EmptyTableRecordCursor;
 import io.questdb.griffin.engine.functions.CursorFunction;
-import io.questdb.std.Misc;
+import io.questdb.griffin.engine.functions.GenericRecordCursorFactory;
 import io.questdb.std.ObjList;
-import io.questdb.std.str.Path;
 
 public class TypeCatalogueFunctionFactory implements FunctionFactory {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/functions/catalogue/TypeCatalogueFunctionFactory.java:34]: check if this factory overrides `isCursor()`

isCursor must return `true` for all factories that return cursor instance

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

**Review by tuanggo — APPROVED** [review]

**Review by bluestreak01 — APPROVED** [review]

**Pull request merged**